text/0014-is-authorized-batch.md

```diff
@@ -0,0 +1,247 @@
+there was support from @aaronjeline and @andrewmwells-amazon for exposing this
+API in Cedar core so that it could be picked up more easily by other language
+bindings, as they expect those bindings to have similar needs and performance
+concerns.
+
+## Detailed design
+
+### Accept a batch of requests with the current shape
+
+This approach creates a new `is_authorized_batch` function, similar to the
+`is_authorized` function, that accepts a list of `Request` structs
+([ref](https://github.com/cedar-policy/cedar/blob/d9201161d8ce45e015c8bcb1cd4191c11039709a/cedar-policy/src/api.rs#L2006))
+and returns a list of `Response` structs
+([ref](https://github.com/cedar-policy/cedar/blob/d9201161d8ce45e015c8bcb1cd4191c11039709a/cedar-policy/src/api.rs#L338)),
+which might have an implementation something like this:
+
+```rust
+impl Authorizer {
+    /// Returns a vector of authorization responses for the input vector of requests
+    /// with respect to the given `PolicySet` and `Entities`.
+    ///
+    /// The order of responses matches the order of requests.
+    pub fn is_authorized_batch(
+        &self,
+        requests: &Vec<Request>,
+        policy_set: &PolicySet,
+        entities: &Entities,
+    ) -> Vec<Response> {
+        let mut responses: Vec<Response> = vec![];
+
+        for request in requests {
+            let response = self.is_authorized(request, policy_set, entities);
+            responses.push(response);
+        }
+
+        responses
+    }
+}
+```
+
+In the special case where the input `requests` list is empty, the function
+should be a no-op and return an empty list.
+
+An advantage of this alternative is that it gives full flexibility to the API
+consumer in how they structure and batch requests.
+
+A disadvantage of this alternative is that for the initial motivating case of a
+`list` operation on a set of resources the principal, action, and context will
+be duplicated across the list of requests. However, this is not necessarily a
+fatal disadvantage, as a language binding author could write optimized versions
+of the binding that translates the special cases on the Rust side of the binding
+and generates the list of `Request` objects from a more-optimized intermediate
+representation.
+
+## Drawbacks
+
+1. Expanding the API surface can be confusing to consumers, as they may not know
+   which of the `is_authorized_*` methods best meets their needs.
+
+2. If this proposal is implemented too conservatively by simply duplicating the
+   `is_authorized` function and dropping in some code to iterate over the input
+   requests, having both `is_authorized` and `is_authorized_batch` APIs could
+   result in double the amount of code to maintain.
+
+3. If this proposal is implemented too aggressively by making `is_authorized`
+   accept a batch of requests and not maintaining the existing function
+   signature, existing Cedar applications would need to be updated.
+
+## Alternatives
```

Comment:
I'd like to propose an additional alternative for discussion.

One drawback with the RFC's proposal as written is that it has to make a single choice as to which parts of the batch request remain constant across all requests, and which parts vary per request.
- Only the resource varies, and principal/action/context/entities/policies are constant?  This seems likely for the filtering-photos use case
- Principal-resource pairs, with action/context/entities/policies constant?
- Principal-action-resource-context tuples, with entities/policies constant?  This might be the most natural interpretation of "batch API"
- Etc

The batch API would be inherently inflexible; it has to basically pick one of these choices, and be suboptimal for the others.

Instead, the various language bindings could expose what I'd call a multi-step interface for `is_authorized()`, similar to what's possible today for Rust users of `is_authorized()`.  One FFI call would construct a Rust `PolicySet` object and return some kind of opaque reference to it.  Another call would construct a Rust `Entities` object and return an opaque reference.  Then the bindings' `is_authorized()` would take these preconstructed opaque references representing policies and entities, rather than re-parsing policies and entities with every call.

The penalty for one-off `is_authorized()` calls is small -- just two extra trips across the FFI boundary, which I imagine will be dwarfed by the actual parsing time for policies and entities in most cases.  The benefit we get is that callers can reuse parsed policies and entities across arbitrary calls.  This API doesn't have to choose between the above options but can easily accommodate them all, as callers choose which arguments remain constant and which vary.  It also avoids the concerns in one of the above threads about the right format to return 100 API results, as in this alternative, each call just returns one result.

Reply:
@cdisselkoen This is an interesting alternative. I'd be curious about lifecycle management for these opaque references; how would the Rust objects that are referenced be freed once the references are no longer used on the other side of the boundary? I'm going to guess that there's a standard pattern for this?

There's a task to gather more metrics to help us understand where time is being spent in the Python binding (https://github.com/k9securityio/cedar-py/issues/14); hopefully this data will help guide us in resolving your comment "The penalty for one-off is_authorized() calls is small -- just two extra trips across the FFI boundary, which I imagine will be dwarfed by the actual parsing time for policies and entities in most cases."

Does this comment match up with @andrewmwells-amazon 's [comment about only constructing the Evaluator once](https://github.com/cedar-policy/rfcs/pull/14#discussion_r1259943823) or does it expand on it?

Reply:
I'd say the two comments are orthogonal.  Constructing the evaluator is different from parsing policies or entities.